app/src/main/java/com/zulip/android/activities/LoginActivity.java

```diff
@@ -182,58 +128,29 @@ public void openHome() {
     }
 
     @Override
-    public void onConnectionFailed(ConnectionResult result) {
-        if (connectionProgressDialog.isShowing()) {
-            // The user clicked the sign-in button already. Start to resolve
-            // connection errors. Wait until onConnected() to dismiss the
-            // connection dialog.
-            if (result.hasResolution()) {
-                try {
-                    result.startResolutionForResult(this, REQUEST_CODE_RESOLVE_ERR);
-                } catch (SendIntentException e) {
-                    // Yeah, no idea what to do here.
-                    connectionProgressDialog.dismiss();
-                    Toast.makeText(LoginActivity.this, R.string.google_app_login_failed, Toast.LENGTH_SHORT).show();
-                }
-            } else {
-                connectionProgressDialog.dismiss();
-                Toast.makeText(LoginActivity.this, R.string.google_app_login_failed, Toast.LENGTH_SHORT).show();
-            }
-        }
-    }
-
-    public void setupGoogleSignIn() {
-        if (mGoogleApiClient == null) {
-            GoogleSignInOptions googleSignInOptions = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
-                    .requestEmail()
-                    .requestIdToken(BuildConfig.GOOGLE_CLIENT_ID)
-                    .build();
-
-            mGoogleApiClient = new GoogleApiClient.Builder(LoginActivity.this)
-                    .addApi(Auth.GOOGLE_SIGN_IN_API, googleSignInOptions)
-                    .addOnConnectionFailedListener(LoginActivity.this)
-                    .build();
-
-            mGoogleApiClient.connect();
-            allowUserToPickAccount();
-        } else {
-            allowUserToPickAccount();
-        }
-
-    }
-
-    private void allowUserToPickAccount() {
-        Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(mGoogleApiClient);
-        startActivityForResult(signInIntent, REQUEST_CODE_SIGN_IN);
-    }
-
-    @Override
     public void onClick(View v) {
         switch (v.getId()) {
             case R.id.google_sign_in_button:
                 connectionProgressDialog.show();
                 saveServerURL();
-                setupGoogleSignIn();
+                AsyncFetchGoogleID asyncFetchGoogleID = new AsyncFetchGoogleID(ZulipApp.get());
+                asyncFetchGoogleID.setCallback(new AsyncTaskCompleteListener() {
+                    @Override
+                    public void onTaskComplete(String result, JSONObject jsonObject) {
+                        try {
+                            JSONObject jsonObject1 = new JSONObject(result);
+                            setupSignIN(jsonObject1.getString("google_client_id"));
```

Comment:
the signature for this method already returns the parsed JSONObject for the result.  You can use that rather than creating a new one.